File: dart/lib/src/packets/src/server/as.dart
part of '../../packets.dart';

class AsPacket extends ServerPacket {
  /// [AsPacket] is the authentication success package.
  ///
  /// This package is part of the package sent from the server to the device.
  AsPacket() : super();

  /// [fromPacket] creates a [AsPacket] from a string package in the format of `Layrz Protocol v2`.
  static AsPacket fromPacket(String raw) {
    if (!raw.startsWith('<As>') || !raw.endsWith('</As>')) {
      throw ParseException('Invalid identification package, should be <As>...</As>');
    }

    final parts = raw.substring(4, raw.length - 5).split(';');

    if (parts.length != 2) {
      throw MalformedException('Invalid package parts, should have 2 parts');
    }

    int? receivedCrc = int.tryParse(parts[1], radix: 16);
    int? calculatedCrc = calculateCrc("${parts.sublist(0, 1).join(';')};".codeUnits);

    if (receivedCrc != calculatedCrc) {
      throw CrcException('Invalid CRC, received: $receivedCrc, calculated: $calculatedCrc');
    }

    return AsPacket();
  }

  /// [toPacket] returns the package in the format of `Layrz Protocol v2`.
  @override
  String toPacket() {
    String payload = ';';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();
    return '<As>$payload$crc</As>';
  }

  @override
  String toString() => toPacket();

  @override
  AsPacket copyWith() {
    return AsPacket();
  }
}

File: go/packets/server/as.go
package server

import (
	"errors"
	"fmt"
	"strconv"
	"strings"

	"github.com/goldenm-software/layrz-protocol/go/v3/internal/wire"
)

// Based on the `Layrz Protocol v2` specification. AsPacket is the success authentication packet
// sent from the server to the device
type AsPacket struct{}

// FromPacket is a method that converts a raw packet to a AsPacket
// based on the `Layrz Protocol v2` specification
//
// Returns an error if the packet is invalid
func (p *AsPacket) FromPacket(raw *string) error {
	if !strings.HasPrefix(*raw, "<As>") || !strings.HasSuffix(*raw, "</As>") {
		return errors.New("invalid package, should be <As>...</As>")
	}

	*raw = strings.TrimPrefix(*raw, "<As>")
	*raw = strings.TrimSuffix(*raw, "</As>")

	parts := strings.Split(*raw, ";")
	if len(parts) != 2 {
		return errors.New("invalid package, should contain 2 parts")
	}

	receivedCrc, err := strconv.ParseUint(parts[1], 16, 16)
	if err != nil {
		return errors.New("cannot convert CRC to integer")
	}

	calculatedCrc := wire.Calculate([]byte(";"))

	if calculatedCrc != uint16(receivedCrc) {
		return fmt.Errorf("invalid CRC, received: %04X, calculated: %04X", receivedCrc, calculatedCrc)
	}

	return nil
}

// ToPacket is a method that converts a AsPacket to a raw packet
// based on the `Layrz Protocol v2` specification
func (p *AsPacket) ToPacket() *string {
	crc := wire.Calculate([]byte(";"))
	content := fmt.Sprintf("<As>;%04X</As>", crc)
	return &content
}
